### PR TITLE
Homework 2

### PR DESCRIPTION
**Pull request by shadston**
Authors: @TalNovatzky, @Shadston

### CODE REVIEW

File: shm/Alcohol.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#include "Alcohol.h"
+
+#include <algorithm>
+
+Alcohol::Alcohol(std::string name, size_t amount, size_t base_price, Time* time, size_t power)
+    : Cargo(name, amount, base_price, time), power_(power) {}
+
+//override from Cargo
```
Comment (nauka-programowania-MA) [inline shm/Alcohol.cpp:8]: Only once in .hpp

File: shm/Alcohol.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+std::string Cargo::getName() const {
+    return name_;
+}
+
+//override from Cargo
```
Comment (nauka-programowania-MA) [inline shm/Alcohol.cpp:13]: Same here. I will not point it further. But remember to fix all.

File: shm/Alcohol.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    }
+}
+
+//override from Cargo
+bool Alcohol::operator==(Cargo& cargo) const {
+    Alcohol* otherAlcohol = dynamic_cast<Alcohol*>(std::addressof(cargo));
+
+    if (!otherAlcohol) {
+        return false;
+    }
```
Comment (nauka-programowania-MA) [inline shm/Alcohol.cpp:41]: or typeid

File: shm/Alcohol.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    ~Alcohol() override;
+
+    std::string getName() const override;
+    size_t getAmount() const override;
+    size_t getPrice() const override;
+    size_t getBasePrice() const override;
+    void nextDay() override;
+    std::string getDescription() const override;
+
+    bool operator==(Cargo& cargo) const override;
```
Comment (nauka-programowania-MA) [inline shm/Alcohol.h:25]: here leave comment override from...

File: shm/Cargo.cpp [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+      time_(time) {
+    time_->subscribe(this);
+}
+
+Cargo::~Cargo() {
+    time_->unsubscribe(this);
+}
+
+void Cargo::update() {
+    this->nextDay();
```
Comment (nauka-programowania-MA) [inline shm/Cargo.cpp:16]: this redundant

File: shm/Fruit.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    time_elapsed_++;
+}
+
+//override from Cargo
+bool Fruit::operator==(Cargo& cargo) const {
+    Fruit* otherFruit = dynamic_cast<Fruit*>(std::addressof(cargo));
+
+    if (!otherFruit) {
+        return false;
+    }
```
Comment (nauka-programowania-MA) [inline shm/Fruit.cpp:42]: or typeid

File: shm/Fruit.cpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+//override from Cargo
+std::string Fruit::getDescription() const {
+    size_t time_to_rot;
+    if (this->expiry_date_ < this->time_elapsed_) {
+        time_to_rot = 0;
+    } else {
+        time_to_rot = this->expiry_date_ - this->time_elapsed_;
+    }
+    return "Name: " + this->name_ + " Time to rot: " + std::to_string(time_to_rot) + " Amount: " + std::to_string(this->amount_) + " Price: " + std::to_string(this->getPrice());
```
Comment (nauka-programowania-MA) [inline shm/Fruit.cpp:73]: this redundant

File: shm/Game.cpp [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+#include "Time.h"
+
+Game::Game(size_t start_money, size_t game_days, size_t final_goal)
+    : start_money_(start_money), game_days_(game_days), final_goal_(final_goal) {}
+
+void Game::startGame() {
+    Time time;
+    Ship ship(10, 3, 3, "Titanic", 0, &time);
+    Player player(ship, start_money_);
+    Map map(&time);
```
Comment (nauka-programowania-MA) [inline shm/Game.cpp:15]: this will be hard to test. You will have problem with replace time in test.

File: shm/Island.cpp [hunk 9]
```diff
@@ -0,0 +1,15 @@
+#include "Island.h"
+
+Island::Island(size_t position_x, size_t position_y, Time* time) {
+    Coordinates coord(position_x, position_y);
+    position_ = coord;
+    store_ = new Store(time);
```
Comment (nauka-programowania-MA) [inline shm/Island.cpp:6]: no raw_tr. Only unique_ptr or shared_ptr

File: shm/Island.h [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+class Island {
+public:
+    Island() = default;
+    Island(size_t position_x, size_t position_y, Time* time);
+    Coordinates getCoordinates() const;
+    Store* getStore();
+
+private:
+    Store* store_;
```
Comment (nauka-programowania-MA) [inline shm/Island.h:15]: shared or unique ptr

File: shm/Island.cpp [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+Store* Island::getStore() {
+    return store_;
```
Comment (nauka-programowania-MA) [inline shm/Island.cpp:10]: when you create smart_pointer call here .get()

File: shm/Item.cpp [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+//override from Cargo
+size_t Item::getPrice() const {
+    return basePrice_ * static_cast<int>(rarity_);
+}
+
+//override from Cargo
+void Item::nextDay() {
+    // it is suppossed to be empty, no mistake here
```
Comment (nauka-programowania-MA) [inline shm/Item.cpp:28]: leve comment like /* not implememnt */

File: shm/Item.cpp [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+//override from Cargo
+bool Item::operator==(Cargo& cargo) const {
+    Item* otherItem = dynamic_cast<Item*>(std::addressof(cargo));
+
+    if (!otherItem) {
+        return false;
+    }
```
Comment (nauka-programowania-MA) [inline shm/Item.cpp:37]: or typeid

File: shm/Observer.h [hunk 14]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+class Observer {
+public:
+    virtual ~Observer(){};
+    virtual void update() = 0;
+};
```
Comment (nauka-programowania-MA) [inline shm/Observer.h:7]: Not as standalone class. You can have more Observer in future and have conflict with names

File: shm/Player.cpp [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+size_t Player::getSpeed() const {
+    return ship_->getSpeed();
+}
+
+Cargo* Player::getCargo(size_t index) const {
+    if (ship_) {
+        return ship_->getCargo(index).get();
+    }
```
Comment (nauka-programowania-MA) [inline shm/Player.cpp:31]: validate index

File: shm/Ship.cpp [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+std::string Ship::getName() const {
+    return name_;
+}
+
+size_t Ship::getId() const {
+    return id_;
+}
+
+void Ship::load(std::shared_ptr<Cargo> cargo) {
+    cargos_.push_back(cargo);
```
Comment (nauka-programowania-MA) [inline shm/Ship.cpp:64]: What if you have the same cargo? Maybe try to accumulate it.

File: shm/Ship.cpp [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+void Ship::unload(Cargo* cargo) {
+    cargos_.erase(std::remove_if(cargos_.begin(), cargos_.end(), [cargo](auto listedCargo) {
+        return *cargo == *listedCargo;
```
Comment (nauka-programowania-MA) [inline shm/Ship.cpp:69]: What if you not sell whole cargo? But only part of it?

**Review by nauka-programowania-MA — CHANGES_REQUESTED** [review]

File: shm/Alcohol.cpp [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+//override from Cargo
+size_t Cargo::getBasePrice() const {
+    return basePrice_;
+}
+
+//override from Cargo
+size_t Alcohol::getPrice() const {
+    return basePrice_ * (power_ / max_power);
```
Comment (WBigaj) [inline shm/Alcohol.cpp:25]: All this element are integer type. That means that division will also be integer type. For example if you have power less than max power(like 90) division will give you 0 and price will become 0.

File: shm/Store.h [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    {"Rum", 30},
+    {"Ale", 8},
+    {"Beer", 3},
+};
+
+std::vector<FruitData> available_fruits = {
+    {"Banana", 10},
+    {"Apple", 5},
+    {"Watermelon", 3},
+};
```
Comment (WBigaj) [inline shm/Store.h:47]: Since those three vectors are predetermined couldn't they be constexpr.

**Review by WBigaj — COMMENTED** [review]
I have checked your code and feft my feedback.

File: shm/Player.cpp [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+
+size_t Player::getAvailableSpace() const {
+    return available_space_;
+}
+
+size_t Player::getMoney() const {
+    return money_;
+}
+
+void Player::setMoney(size_t amount) {
```
Comment (majkel84) [inline shm/Player.cpp:16]: just a suggestion, try to keep same names for the variables which set class members
```suggestion
void Player::setMoney(size_t money) {
```

**Review by majkel84 — APPROVED** [review]

File: shm/Ship.cpp [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+        crew_ -= num;
+    }
+    return *this;
+}
+
+Ship& Ship::operator+=(size_t num) {
+    crew_ += num;
+    if (crew_ > max_crew_) {
+        crew_ = max_crew_;
+    }
```
Comment (WBigaj) [inline shm/Ship.cpp:32]: I think it's mistake to add before checking. If you add value that exceeds size_t maximum value you may get garbage value that's lower than crew limit.

**Review by WBigaj — COMMENTED** [review]
When I tested my SHM project i found error in ship operator overloading,which I have seen in many project. I have found other type error,when I checked your code again.

**Comment by nauka-programowania-MA** [comment]
{
"requirement_ids": [2568, 2568, 2568, 2571, 2571, 2571, 2570],
"points": [5, 5, 5, 8, 8, 8, 6],
"users": ["TalNovatzky", "shadston"]
}

File: shm/Store.h (hunk 19 above)
Comment (MaciejSalamonski) [reply shm/Store.h:47]: +1

**Pull request closed without merging**